lib/aiken-design-patterns/linked-list/ordered.ak

```diff
@@ -0,0 +1,251 @@
+use aiken/collection/dict.{Dict}
+use aiken/collection/list
+use aiken/primitive/bytearray
+use aiken_design_patterns/linked_list/types.{Empty, Key, NodeDatum, NodeKey}
+use aiken_design_patterns/linked_list/unordered
+use aiken_design_patterns/linked_list/utils.{
+  get_key, get_link, get_raw_key_and_link,
+}
+use cardano/address.{Address}
+use cardano/assets.{AssetName, flatten}
+use cardano/transaction.{InlineDatum, Input, Output}
+
+pub fn list_state_transition(node_mint: Dict<AssetName, Int>) {
+  unordered.list_state_transition(node_mint)
+}
+
+pub fn modify_data(
+  node_mint: Dict<AssetName, Int>,
+  own_input: Input,
+  own_output: Output,
+  node_nft_policy_id: ByteArray,
+  node_nft_asset_name: ByteArray,
+) -> Bool {
+  unordered.modify_data(
+    node_mint,
+    own_input,
+    own_output,
+    node_nft_policy_id,
+    node_nft_asset_name,
+  )
+}
+
+pub fn is_root_node(node_key: NodeKey) {
+  unordered.is_root_node(node_key)
+}
+
+pub fn is_last_node(node_link: NodeKey) {
+  unordered.is_last_node(node_link)
+}
+
+pub fn is_empty_list(node_key: NodeKey, node_link: NodeKey) {
+  unordered.is_empty_list(node_key, node_link)
+}
+
+pub fn key_added(key: NodeKey, node_mint: Dict<AssetName, Int>) {
+  unordered.key_added(key, node_mint)
+}
+
+pub fn key_removed(key: NodeKey, node_mint: Dict<AssetName, Int>) {
+  unordered.key_removed(key, node_mint)
+}
+
+pub fn is_member(
+  k: NodeKey,
+  node_key: NodeKey,
+  node_address: Address,
+  linked_list_address: Address,
+) {
+  unordered.is_member(k, node_key, node_address, linked_list_address)
+}
+
+pub fn is_not_member(
+  k: NodeKey,
+  node_key: NodeKey,
+  node_link: NodeKey,
+  node_address: Address,
+  linked_list_address: Address,
+) {
+  or {
+    node_address != linked_list_address,
+    and {
+      k != Empty,
+      {
+        expect Key(actual_key) = k
+        and {
+          or {
+            unordered.is_root_node(node_key),
+            {
+              expect Key(actual_node_key) = node_key
+              bytearray.compare(actual_node_key, actual_key) == Less
+            },
+          },
+          or {
+            unordered.is_last_node(node_link),
+            {
+              expect Key(actual_node_link) = node_link
+              bytearray.compare(actual_key, actual_node_link) == Less
+            },
+          },
+        }
+      },
+    },
+  }
+}
+
+//Application code must ensure that this action can happen only once
+pub fn init(
+  node_outputs: List<Output>,
+  node_mint: Dict<AssetName, Int>,
+  node_cs: ByteArray,
+) {
+  unordered.init(node_outputs, node_mint, node_cs)
+}
+
+pub fn deinit(
+  node_in_outputs: List<Output>,
+  node_mint: Dict<AssetName, Int>,
+  node_cs: ByteArray,
+) {
+  unordered.deinit(node_in_outputs, node_mint, node_cs)
+}
+
+pub fn prepend(
+  key_to_prepend: NodeKey,
+  prepended_node_output_index: Int,
+  anchor_node_output_index: Int,
+  node_in_outputs: List<Output>,
```

Comment:
I find the `in_outputs` suffix very confusing. Half the time, it makes me think of `node_outputs`.
We should just call them `node_inputs`.

IMO it's not a big deal, as we already have `node_mint : Dict<AssetName, Int>` instead of the usual `Value`.

When we're actually dealing with a `List<Input>`, we can suffix it with `inputs_with_ids`.

```suggestion
  node_inputs: List<Output>,
```

Reply:
Resolved by https://github.com/Anastasia-Labs/aiken-design-patterns/pull/12